src/runtime/storageNG/reference-mode-store.ts

```diff
@@ -0,0 +1,598 @@
+   * should not return until the response relevant to the message has been received.
+   * 
+   * When handling proxy messages, this implies 2 rounds of update - first the backing
+   * store needs to be updated, and once that has completed then the container store needs
+   * to be updated.
+   */
+  async onContainerStore(message: ProxyMessage<ReferenceContainer>) {
+    return this.enqueue({from: 'container', message});
+  }
+
+  async onBackingStore(message: ProxyMessage<CRDTEntityTypeRecord<S, C>>, muxId: string) {
+    return this.enqueue({from: 'backing', message, muxId});
+  }
+
+  async onProxyMessage(message: ProxyMessage<Container>): Promise<boolean> {
+    return this.enqueue({from: 'proxy', message});
+  }
+
+  /**
+   * enqueue an incoming update onto the object-wide queue and return a promise that will be resolved
+   * when the update is processed.
+   */
+  private async enqueue(entry: PreEnqueuedMessage<Container, CRDTEntityTypeRecord<S, C>, ReferenceContainer>): Promise<boolean> {
+    return new Promise((resolve, reject) => {
+      const startProcessing = this.receiveQueue.length === 0;
+      this.receiveQueue.push({...entry, promise: resolve});
+      if (startProcessing) {
+        void this.processQueue();
+      }
+    });
+  }
+
+  private async processQueue() {
+    while (this.receiveQueue.length > 0) {
+      const nextMessage = this.receiveQueue[0];
+      console.log('pQ processing', nextMessage);
+      switch (nextMessage.from) {
+        case 'proxy':
+          nextMessage.promise(await this.handleProxyMessage(nextMessage.message));
+          break;
+        case 'backing':
+          nextMessage.promise(await this.handleBackingStore(nextMessage.message, nextMessage.muxId));
+          break;
+        case 'container':
+          nextMessage.promise(await this.handleContainerStore(nextMessage.message));
+          break;
+        default:
+          throw new Error('invalid message type');
+      }
+      this.receiveQueue.shift();
+    }
+  }
+
+  /**
+   * Handle an update from the container store.
+   * 
+   * Operations and Models either enqueue an immediate send (if all referenced entities
+   * are available in the backing store) or enqueue a blocked send (if some referenced
+   * entities are not yet present).
+   * 
+   * Note that the blocking mechanism isn't version-aware, so removes followed by
+   * adds may not correctly sync. If this turns out to be a problem then we can
+   * add version information to references and update the blocking store to gate
+   * on a version as well as presence.
+   * 
+   * Sync requests are propagated upwards to the storage proxy.
+   */
+  private async handleContainerStore(message: ProxyMessage<ReferenceContainer>) {
+    switch (message.type) {
+      case ProxyMessageType.Operations: {
+        for (const operation of message.operations) {
+          const reference = this.operationElement<Reference>(operation);
+          let getEntity;
```

Comment:
Can you add a type for this? I think it might be: `let getEntity: () => (Entity | null);`